internal/policy/policy.go

```diff
@@ -170,10 +172,38 @@ func LoadStateForEntry(ctx context.Context, repo *git.Repository, e rsl.EntryTyp
 
 		if entry.Name == fmt.Sprintf("%s.json", RootRoleName) {
 			state.RootEnvelope = env
+			rootMetadata, err := state.GetRootMetadata()
+			if err != nil {
+				return nil, err
+			}
+			for _, key := range rootMetadata.Keys {
+				key := key // avoid iterator pointer issues
+				key.ID()   //nolint:errcheck
+				state.AllPublicKeys = append(state.AllPublicKeys, key)
+			}
 		} else if entry.Name == fmt.Sprintf("%s.json", TargetsRoleName) {
 			state.TargetsEnvelope = env
+			targetsMetadata, err := state.GetTargetsMetadata(TargetsRoleName)
+			if err != nil {
+				return nil, err
+			}
+			for _, key := range targetsMetadata.Delegations.Keys {
+				key := key // avoid iterator pointer issues
+				key.ID()   //nolint:errcheck
+				state.AllPublicKeys = append(state.AllPublicKeys, key)
+			}
 		} else {
-			state.DelegationEnvelopes[strings.TrimSuffix(entry.Name, ".json")] = env
+			roleName := strings.TrimSuffix(entry.Name, ".json")
+			state.DelegationEnvelopes[roleName] = env
+			targetsMetadata, err := state.GetTargetsMetadata(roleName)
+			if err != nil {
+				return nil, err
+			}
+			for _, key := range targetsMetadata.Delegations.Keys {
+				key := key // avoid iterator pointer issues
+				key.ID()   //nolint:errcheck
+				state.AllPublicKeys = append(state.AllPublicKeys, key)
+			}
```

Comment:
I'd add comments explaining the differences between these 2 (they look very similar).